### PR TITLE
SNOW-982770: [Local Testing] Add support for rank related features.

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-982770

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This PR does a few things to add support for rank related features to local testing.

- Adds some metadata to ColumnEmultors so that they can retain information about how they have been sorted. Arguably this could be made a private variable as I would not expect users to need direct access to this information outside of mocking their own functions.
- Modifies windowed functions to retain calculated expression columns that are used for ordering when passing to a mocked rank function.
- Adds the mocked rank functions and expands test examples to include more branches in the local testing logic.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_snowflake_data_type.py [hunk 1]
```diff
@@ -344,6 +349,7 @@ def __init__(
             {} if not sf_types_by_col_index else sf_types_by_col_index
         )
         self._null_rows_idxs_map = {}
+        self.sorted_by = []
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_snowflake_data_type.py:352]: how is this information used?

File: tests/integ/scala/test_function_suite.py [hunk 2]
```diff
@@ -4484,29 +4484,67 @@ def test_iff(session, local_testing_mode):
...
 
-@pytest.mark.skipif(
-    "config.getoption('local_testing_mode', default=False)",
-    reason="cume_dist is not yet supported in local testing mode.",
-)
 def test_cume_dist(session):
     Utils.check_answer(
         TestData.xyz(session).select(
-            cume_dist().over(Window.partition_by(col("X")).order_by(col("Y")))
+            "X", "Y", cume_dist().over(Window.partition_by(col("X")).order_by(col("Y")))
```
Comment (sfc-gh-aling) [inline tests/integ/scala/test_function_suite.py:4490]: I see some changes in the test, like adding "X", "Y" in the select statement, change sort from `False` to `True` in the check_answer.

what's the purpose of the change here? is the previous test not working in the implementation or it's for adding more test coverage here?

File: src/snowflake/snowpark/mock/_snowflake_data_type.py (hunk 1 above)
Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_snowflake_data_type.py:352]: This metadata allows the `handle_order_by_clause` to annotate a TableEmulator with the columns that it hsa been sorted by. That is useful because some of the rank functions need that information to break ties if the data is sorted.

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/mock/_snowflake_data_type.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_snowflake_data_type.py:352]: oh yes, I saw that piece of information used by the _rank method

File: tests/integ/scala/test_function_suite.py (hunk 2 above)
Comment (sfc-gh-aling) [reply tests/integ/scala/test_function_suite.py:4490]: thanks, now I understand why cols X and Y are introduced.

btw does the server guarantee the same ordering or can return different orders. wondering if we should mention this in the limitation or known gap section of local testing framework.

**Review by sfc-gh-aling — APPROVED** [review]

File: tests/integ/scala/test_function_suite.py (hunk 2 above)
Comment (sfc-gh-jrose) [reply tests/integ/scala/test_function_suite.py:4490]: I believe the server order is consistent, not sure if it has a guarentee.
